### ISSUE TITLE
`DeserializeGame` fails on double `GameParameters` that represent integer values

### ISSUE THREAD

**Issue by gabrfarina** [deepmind/open_spiel#382]
Hit as part of implementing the Battleship game (#376).

`DeserializeGame` fails to deserialize games that have (at least) one double game parameter whose value corresponds to an integer. While it's easy to trigger this failure by hand, it is important to note that it will also be triggered automatically by `TestSerializeDeserialize`, which automatically tries to serialize and then deserialize a game --- in fact, that's how I hit the issue.

The issue is due to the fact that the parser for game parameters (snippet below) is *too eager* to match integers: any sequence composed uniquely of characters "+-0123456789" is immediately parsed as an integer.
https://github.com/deepmind/open_spiel/blob/91da643188fcacd7a57bf5a64ee392d6c919dcfc/open_spiel/game_parameters.cc#L168-L181

Unfortunately, it is not clear to me how to fix this issue:
* Asking that all double values always include either 'e' or a period '.' is very non-user-friendly, and would still require changes in the serialization code, which currently serializes a double like "2.0" to just "2". I think that imposing a restriction on how doubles are serialized is going to impose unnecessary burden to users.
* On the other hand, it seems like the parsing of game parameters is not done taking into account the desired resulting types. The type of the game parameters is determined lexically at game deserialization time, and the deserialized parameters are them matched against the expected schema by the `ValidateParams` method.

I can see two ways to fix the issue:
1. Either have `DeserializeGame` and `GameParameterFromString` have access to information about the schema of game parameters they should parse. This seems the cleanest solution, but would require some heavy reworking of some internal pipelines from what I can see.
2. Have `ValidateParams` include some failure recovery strategy that will promote integers to doubles instead of failing. This has the advantage that the fix would be contained to a single method. But it seems very hacky: for example, if a string parameter has the misfortune of being the string `"true"`, the parameter would currently be parsed as `bool`, even if the schema says that it should be string.

I think as a first step towards solving this, the maintainers should agree on what solution strategy to adopt. I am happy to help with whatever you might determine is best.

This issue is currently a blocker for the implementation of the Battleship game, which includes a double game parameter whose value will often be integer --- as is 2.0 by default.

**Comment by gabrfarina** [comment]
A couple more thoughts about why option 2 above is very hacky:
* The current parser fails to parse values represented in exponential notation. For example, "1e2" would be parsed as a string instead of the double value "100.0". So, if we were to go with option 2, we would need to consider a promotion path from string to double in some cases.
* The promotion paths might be lossy. Suppose that we fix the point above and we parse a parameter with value "1e2" as double, only to find out later in `ValidateParams` that the schema says the parameter was actually supposed to be a string. How do we recover from this? Do we promote from double to string again, and store the string "100" at this point? The original string was "1e2" and would not be parsed as double, the promoted back to string, and stored as the completely different string "100". Probably too surprising to an end user. This concern applies even if the string was "100.0", and would now be promoted back to "100" losing the final ".0" characters.

**Comment by lanctot** [comment]
Oh yes, this seems like a tough one.

For now, does setting the parameter value to 2.0 (even if it's not user friendly) resolve the problem for Battleship? If so, let's adopt that solution for now because I imagine this will take some time to resolve.

**Comment by gabrfarina** [comment]
Yes, I think a solution would be to just change the way double parameters are serialized by default, so that `SerializeDeserializeTest` passes.

I am happy to submit a quick PR for that later today and then rebase the battleship branch once it lands.

Do you have any preference for how to force a dot in the serialization of a double? The standard library way is probably to use std::fixed and std::setprecision from iomanip? Or perhaps abseil offers better formatting options for numbers? Or if you like it better there's always sprintf.

**Comment by lanctot** [comment]
Nope no preference, choose whatever works. Appreciate the fixes, thanks. If you submit a PR this weekend I will make sure to import it to make Monday's update so it unblocks Battleship asap.

**Comment by elkhrt** [comment]
Would it work to have GetValue<double> return the int value if the double
isn't found?

**Comment by gabrfarina** [comment]
@elkhrt I suspect the check that fails in `ValidateParams` compares the type of the param (using `.Type()`) against the schema, so I think that changing `.GetValue()` alone would not fix the issue directly.

You could in theory have what you were saying, plus tweak `ValidateParam` so that it got an int instead of a double, it will not throw.

This effectively would create a promotion path from int to double, though we would need to be careful to make sure that no other code throws if the parameter has type kInt but the code was expecting kDouble if we want to be consistent.

**Comment by gabrfarina** [comment]
I have fixed locally the double problem as discussed with @lanctot. However, another problem related to this came up: in Battleship I was passing in the configuration of the ship sizes as a comma separated list of integers stored as a string. This leads to problems in the serialization / deserialization:
 * (A) When there is only one ship of size `1`, then the string is `"1"`. That gets printed without double quotes, so it gets read as an integer.
 * (B) If there are two ships of size `1` and `2`, respectively, the serialization would include `"1,2"`, and so the parser would think that there are multiple arguments and dies with "Unkonwn parameter".

The only reasonable fix I see to as a bandaid for now is to require that lists of values always start with `[` and end with `]` (to fix problem A) and be `;`-separated or space-separated (instead of comma separated) to avoid problem B.

**Comment by lanctot** [comment]
I say let's do the bandaid for now and rethink this after Battleship is in. I am generally in favor of list params, just worried that it might be a lot of work to do properly and don't want it to keep holding up that PR.

**Comment by gabrfarina** [comment]
I agree 👍 

I'll open a PR that includes some regression tests for the parser in a minute. I also caught another issue where the double value 10^-9 would by default be serialized as the string "1e-9" when using abseil's `StrCat`, but the current `GameParameter` parser does not handle doubles in exponential notation. That's fixed now, I have forced the serialization to not use the exponential notation.

If you preferred the other approach (having the parser handle exponential notation), I'm happy to amend the pull request.
